music/src/core/visualizer.ts

```diff
@@ -456,3 +458,237 @@ export class PianoRollSVGVisualizer extends BaseVisualizer {
+   * adaption to a modified score, modifyied notes previously drawn will be
+   * ignored (you can always `clear()` and `redraw()` for a full redraw). 
+   * Please have in mind `mm.Player` does not have this incremental capability
+   * so, once the player had started, it will go on ignoring the changes. 
+   * 
+   * @param activeNote (Optional) If specified, this `Note` will be painted
+   * in the active color and there won't be an actual redrawing, but a 
+   * re-colouring of the involved note heads, accidentals, dots and ties 
+   * (activated and de-activated ones). Otherwise, all musical symbols which 
+   * were not processed yet will be drawn to complete the score.
+   * @param scrollIntoView (Optional) If specified and the active note to be 
+   * highlited is not visualized in the container DIV, the later will be 
+   * scrolled so that the note is viewed in the right place. This can be 
+   * altered by `AdvancedVisualizerConfig.scrollType`.
+   * @returns The x position of the highlighted active note relative to the 
+   * beginning of the SVG, or -1 if there wasn't any given active note. Useful
+   * for automatically advancing the visualization if needed.
+   */
+  public redraw(
+    activeNote?: NoteSequence.INote, 
+    scrollIntoView?: boolean 
+  ) : number {
+    if (this.drawnNotes !== this.noteSequence.notes.length) {
+      this.render.scoreInfo = this.getScoreInfo(this.noteSequence);
+    }
+    const activeNoteInfo = activeNote ? 
+      this.getNoteInfo(activeNote) : undefined;
+    return this.render.redraw(activeNoteInfo, scrollIntoView);
+  }
+
+  private isNoteInInstruments(note: NoteSequence.INote) : boolean {
+    if (note.instrument === undefined || this.instruments.length === 0) {
+      return true; // No instrument information in note means no filtering
+    }
+    else { // Instrument filtering
+      return this.instruments.indexOf(note.instrument) >= 0;
+    }
+  }
+
+  private timeToQuarters(time: number): number {
+    const q = time * this.noteSequence.tempos[0].qpm / 60;
+    return Math.round(q*16)/16; // Current resolution = 1/16 quarter
+  }
+
+  private quantizedStepsToQuarters(steps: number): number {
+    const q = steps / this.noteSequence.quantizationInfo.stepsPerQuarter;
+    return Math.round(q*16)/16; // Current resolution = 1/16 quarter
+  }
+
+  private getNoteInfo(
+    note: NoteSequence.INote
+  ): sr.NoteInfo {
+    const startQ = this.sequenceIsQuantized ?
+      this.quantizedStepsToQuarters(note.quantizedStartStep):
+      this.timeToQuarters(note.startTime);
+    const endQ = this.sequenceIsQuantized ?
+      this.quantizedStepsToQuarters(note.quantizedEndStep):
+      this.timeToQuarters(note.endTime);
+    return {
+      start: startQ,
+      length: endQ - startQ,
+      pitch: note.pitch,
+      intensity: note.velocity
+    };
+  }
+
+  // TODO: Update when ProtocolBuffer covers a quantizedStep version
+  private getTempoInfo (
```

Comment:
These functions seem to only be used in one function, getScoreInfo (and some only once). I'm not sure they need to be functions at all, private or not.

Reply:
Ok. It was clearer to me, because of the arrow functions in the `forEach` loop, but I can merge them in longer sentences. No problem at all.